infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java

```diff
@@ -0,0 +1,149 @@
+   *
+   * <p>If multiple pods have labels, annotations, additionalProperties with the same key then
+   * override will happen and result pod template will have only the last value. You may need to
+   * reconfigure objects that rely on these value, like selector field in Service. You can use all
+   * of pod template labels for reconfiguring or single {@link #DEFAULT_DEPLOYMENT_NAME} label that
+   * contains deployment name;
+   *
+   * @param podsData the pods data to be merged
+   * @return a new Deployment that have pod template that is result of merging the specified lists
+   * @throws ValidationException if pods can not be merged because of critical names collisions like
+   *     volumes names
+   */
+  public Deployment merge(List<PodData> podsData) throws ValidationException {
+    Deployment baseDeployment = createEmptyDeployment(DEFAULT_DEPLOYMENT_NAME);
+    PodTemplateSpec basePodTemplate = baseDeployment.getSpec().getTemplate();
+    ObjectMeta basePodMeta = basePodTemplate.getMetadata();
+    PodSpec baseSpec = basePodTemplate.getSpec();
+
+    Set<String> containerNames = new HashSet<>();
+    Set<String> initContainerNames = new HashSet<>();
+    Set<String> volumes = new HashSet<>();
+    Set<String> pullSecrets = new HashSet<>();
+    for (PodData podData : podsData) {
+      // if there are entries with such keys then values will be overridden
+      ObjectMeta podMeta = podData.getMetadata();
+      basePodMeta.getLabels().putAll(podMeta.getLabels());
+      basePodMeta.getAnnotations().putAll(podMeta.getAnnotations());
+      basePodMeta.getAdditionalProperties().putAll(podMeta.getAdditionalProperties());
+
+      for (Container container : podData.getSpec().getContainers()) {
+        String containerName = container.getName();
+
+        // generate container name to avoid collisions
+        while (!containerNames.add(container.getName())) {
+          containerName = NameGenerator.generate(container.getName(), 4);
+          container.setName(containerName);
+        }
+
+        // store original recipe machine name
+        basePodMeta
+            .getAnnotations()
+            .put(
+                format(MACHINE_NAME_ANNOTATION_FMT, containerName),
+                Names.machineName(podMeta, container));
+
+        baseSpec.getContainers().add(container);
+      }
+
+      for (Container initContainer : podData.getSpec().getInitContainers()) {
+        // generate container name to avoid collisions
+        while (!initContainerNames.add(initContainer.getName())) {
+          initContainer.setName(NameGenerator.generate(initContainer.getName(), 4));
+        }
+        baseSpec.getInitContainers().add(initContainer);
+      }
+
+      for (Volume volume : podData.getSpec().getVolumes()) {
+        if (!volumes.add(volume.getName())) {
+          throw new ValidationException(
+              format(
+                  "Pods have to have volumes with unique names but there are multiple `%s` volumes",
```

Comment:
Can you elaborate why we can't rewrite volume names with the other volume related objects?

Reply:
We can rename volume with other volume references (like in containers and initContainers) and I'm going to create a separate issue to implement such improvement.
Do you think it is critical for this PR and should be handled here?

Reply:
No, it is not critical on my opinion